`Exam mode`: Add checklist overview

willergeorg opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
#### General
<!-- You only need to choose one of the first two check items: Generally, test on the test servers. -->
<!-- If it's only a small change, testing it locally is acceptable and you may remove the first checkmark. If you are unsure, please test on the test servers. -->
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls..
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/)..
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
This PR introduces a smaller version of the exam checklist as overview on exam management page

### Description
<!-- Describe your changes in detail -->
We decided to replace the current table columns representing a wide range of different dates with some kind of roadmap summarizing the associated exam checklist.
This roadmap is separated in three steps:

1. Exam preparation
2. Exam conduction
3. Exam correction

Each step contains some sub steps that are displayed below its title.

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->
Testing this is quite tricky as we need to check several different states a exam could have.
Basically, every sub step in exam preparation is something that can be sufficiently done or not. Depending on this, it is shown a green check or a red cross next to the sub step.
Below I added a table containing every factor that leads for a specific sub step of exam preparation to be done sufficiently or not. In this table, the different states for the bubbles above the different steps are listed as well.

**Important**: Also check that there are no inconsistencies between the overview and the actual exam checklist (especially when it comes to the amount of submitted and started exams or the complaints)
Prerequisites:
- 1 Instructor
- Several Exams with different states

0. For each possible state, bring an exam to this state, then:
1. Log in to Artemis
2. Navigate to Course Administration
3. Navigate to exam management tab of the course and inspect the overview
4. Check for inconsistencies between overview and exam checklist

![image](https://user-images.githubusercontent.com/80622272/151666020-64f70898-e53b-4c19-b868-dad01d0d5717.png)

### Review Progress
<!-- Each Pull Request should be reviewed by at least two other developers. The code as well as the functionality (= manual test) needs to be reviewed. -->
<!-- The reviewer or author check the following boxes depending on what was reviewed or tested. All boxes should be checked before merge. -->
<!-- You can add additional checkboxes if it makes sense to only review parts of the code or functionality. -->
<!-- When changes are pushed, uncheck the affected boxes. (Not all changes require full re-reviews.) -->

#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage
<!-- Please add the test coverages for all changed files here. You can see this when executing the tests locally (see build.gradle and package.json) or when looking into the corresponding Bamboo build plan. -->
<!-- Please use the schema "Branches % | Lines %" -->
<!-- Lines are the main reference but a significantly lower branch percentage can indicate missing edge cases in the tests. -->
<!-- - ExerciseService.java: 85% | 77% -->
<!-- - programming-exercise.component.ts: 13% | 95% -->
`exam-status.component.ts`: 97.56% | 100%
`exam-checklist.component.ts`: 75% | 100%
`exam-checklist.service.ts`: 84.84% | 100%


### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
![image](https://user-images.githubusercontent.com/80622272/151666863-aa401b58-0e40-4ec2-8ad1-8ff6f48f046a.png)

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.html:
```diff
@@ -0,0 +1,112 @@
+            </div>
+            <ol *ngIf="examChecklist">
+                <li>
+                    <span
+                        >{{ 'artemisApp.examStatus.conduction.startedExams' | artemisTranslate }}:
+                        {{ examChecklist.numberOfExamsStarted }}
+                        ({{ numberOfGeneratedStudentExams > 0 ? round((examChecklist.numberOfExamsStarted * 100) / examChecklist.numberOfGeneratedStudentExams!, 1) : 0 }}%)</span
+                    >
```
```suggestion
                    <span>
                        {{ 'artemisApp.examStatus.conduction.startedExams' | artemisTranslate }}:
                        {{ examChecklist.numberOfExamsStarted }}
                        ({{ numberOfGeneratedStudentExams > 0 ? round((examChecklist.numberOfExamsStarted * 100) / examChecklist.numberOfGeneratedStudentExams!, 1) : 0 }}%)
                    </span>
```

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.html:
```diff
@@ -0,0 +1,112 @@
+                    >
+                </li>
+                <li>
+                    <span
+                        >{{ 'artemisApp.examStatus.conduction.submittedExams' | artemisTranslate }}:
+                        {{ examChecklist.numberOfExamsSubmitted }}
+                        ({{ numberOfGeneratedStudentExams > 0 ? round((examChecklist.numberOfExamsSubmitted * 100) / numberOfGeneratedStudentExams!, 1) : 0 }}%)</span
+                    >
```
```suggestion
                    <span>
                        {{ 'artemisApp.examStatus.conduction.submittedExams' | artemisTranslate }}:
                        {{ examChecklist.numberOfExamsSubmitted }}
                        ({{ numberOfGeneratedStudentExams > 0 ? round((examChecklist.numberOfExamsSubmitted * 100) / numberOfGeneratedStudentExams!, 1) : 0 }}%)
                    </span>
```

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,178 @@
+
+    ngOnChanges(): void {
+        // Step 1: Exam preparation
+        // Step 1.1: Configure exercises
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.CONFIGURE_EXERCISES, this.areAllExercisesConfigured());
+        // Step 1.2: Register students
+        this.configurationStepToFlagMap.set(ExamConfigurationStep.REGISTER_STUDENTS, this.examChecklistService.checkAtLeastOneRegisteredStudent(this.exam));
+        this.examChecklistService.getExamStatistics(this.exam).subscribe((examStats) => {
+            // Step 1.3: Generate student exams
+            this.configurationStepToFlagMap.set(ExamConfigurationStep.GENERATE_STUDENT_EXAMS, this.examChecklistService.checkAllExamsGenerated(this.exam, examStats));
+            // Step 1.4: Prepare exercise start
+            const allExamsGenerated = this.configurationStepToFlagMap.get(ExamConfigurationStep.GENERATE_STUDENT_EXAMS)!;
+            this.configurationStepToFlagMap.set(ExamConfigurationStep.PREPARE_EXERCISE_START, !!examStats.allExamExercisesAllStudentsPrepared && allExamsGenerated);
+            this.examPreparationFinished = this.isExamPreparationFinished();
+            this.examChecklist = examStats;
+            this.numberOfGeneratedStudentExams = this.examChecklist.numberOfGeneratedStudentExams ?? 0;
+        });
```
The other steps are updated in their own method, I think this one should be extracted to one as well.

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,178 @@
+    }
+
+    /**
+     * Auxiliary method that determines if all configuration steps for the exam exercises are finished. There are 5 requirements that have to be fulfilled:
+     * 1. The exam has at least one exercise group
+     * 2. The number of exam exercises equals number of exercise groups
+     * 3. Each exercise group contains at least one exercise
+     * 4. Within each exercise group, exercises have same maximum points
+     * 5. Maximum points of exam are reachable
```
```suggestion
     * Auxiliary method that determines if all configuration steps for the exam exercises are finished.
```
I don’t think the additional list is necessary and neither are the inline comments inside the method, as the method names have been chosen nicely to already explain what is checked.

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,178 @@
+        if (!this.exam.examStudentReviewEnd) {
+            this.examReviewState = ExamReviewState.UNSET;
+        } else if (
+            (!this.exam.examStudentReviewStart && this.exam.examStudentReviewEnd && this.exam.examStudentReviewEnd.isAfter(dayjs())) ||
+            (this.exam.examStudentReviewStart && this.exam.examStudentReviewStart.isBefore(dayjs()) && this.exam.examStudentReviewEnd.isAfter(dayjs()))
```
New method: `isExamRunning(): boolean`

b-fein commented on src/main/webapp/i18n/de/exam.json:
```diff
@@ -458,6 +458,31 @@
+                "examCorrection": "Klausurkorrektur",
+                "correctionRoundProgress": "Korrekturfortschritt in Runde {{ index }}",
+                "examReview": "Klausureinsicht",
+                "unset": "Nicht angesetzt",
```
```suggestion
                "unset": "Nicht gesetzt",
```
I assume this translation is used when some value of the exam is not yet set to a proper value.

willergeorg replied:
The translation is used in order to indicate that the exam review phase is not set by the instructor. It might be personal preference but I think in this case "angesetzt" fits a bit better than "gesetzt". What do you think about this?
![image](https://user-images.githubusercontent.com/80622272/152125278-dfcd0da0-6b6e-4dd6-be42-52dc99039a13.png)

b-fein replied:
Then I assumed the wrong context. In this case the current translation definitely makes more sense.

b-fein commented on src/main/webapp/i18n/en/exam.json:
```diff
@@ -458,6 +458,31 @@
+            },
+            "correction": {
+                "examCorrection": "Exam correction",
+                "correctionRoundProgress": "Correction progress round {{ index }}",
```
```suggestion
                "correctionRoundProgress": "Correction progress in round {{ index }}",
```

b-fein reviewed (approved):
Code review: only minor comments, looks good! :+1:

willergeorg commented:
> Code review: only minor comments, looks good! 👍

Except the change from "angesetzt" to "gesetzt" I incorporated your feedback in fc38a75, 5c0c4ae, de2fa07 and 34b4f76.
Regarding the list in the method documentation of `areAllExercisesConfigured()`: I removed it as you suggested because I had to admit that the method is not very complex, the names are mostly self-explanatory and the specific method documentations should do the rest. In general I have the feeling that we are a bit economical when it comes to in-depth method documentation. This often leads to code that is hard to maintain and modify for any developer other than the original author. I would prefer skipping a documentation comment for a method that seems to be trivial for me rather than spending quite some time in understanding legacy logic that is not sufficiently documented because the author assumed it to be trivial.

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,187 @@
+     * @private
+     */
+    private examAlreadyStarted(): boolean {
+        return this.exam.startDate! && this.exam.startDate!.isBefore(dayjs());
```
```suggestion
        return this.exam.startDate && this.exam.startDate.isBefore(dayjs());
```
The check if the `startDate` is defined shouldn’t require the `!` on the left side. On the right side it should not be needed because the left side already checked that is defined.

willergeorg replied:
This was my perception as well. Unfortunately, the compiler has other ideas. 😅 I would prefer omitting `undefined` as return type. What do you think?
![image](https://user-images.githubusercontent.com/80622272/152150750-1fae1b0d-1cc3-499e-87db-1853b9cc4dac.png)

b-fein replied:
Interesting, well if the compiler wants to have the `!` then so they should definitely stay to avoid the `| undefined` as return type.

(Would `this.exam.startDate?.isBefore(dayjs()) ?? false` work? I’m not sure if that would improve the code or if the current version is better.)

willergeorg replied:
Good catch, this one would work. But I share your concerns that this might make the code a bit more complicated than it is now.
In any case, I think it is sufficient if we keep the exclamation sign for the first `startDate` and remove the second one.

willergeorg replied:
I removed the exclamation marks in 93c3171

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,187 @@
+     * @private
+     */
+    private examAlreadyEnded(): boolean {
+        return this.exam.endDate! && this.exam.endDate!.isBefore(dayjs());
```
```suggestion
        return this.exam.endDate && this.exam.endDate.isBefore(dayjs());
```
As above.

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,187 @@
+     * @returns boolean indicating whether configuration is finished
+     */
+    private areAllExercisesConfigured(): boolean {
+        const atLeastOneGroup = this.examChecklistService.checkAtLeastOneExerciseGroup(this.exam);
+
+        const numberOfExercisesEqual = this.examChecklistService.checkNumberOfExerciseGroups(this.exam);
+
+        const noEmptyExerciseGroup = this.examChecklistService.checkEachGroupContainsExercise(this.exam);
+
+        const maximumPointsEqual = this.examChecklistService.checkPointsExercisesEqual(this.exam);
+
+        const examPointsReachable = this.examChecklistService.checkTotalPointsMandatory(maximumPointsEqual, this.exam);
```
```suggestion
        const atLeastOneGroup = this.examChecklistService.checkAtLeastOneExerciseGroup(this.exam);
        const numberOfExercisesEqual = this.examChecklistService.checkNumberOfExerciseGroups(this.exam);
        const noEmptyExerciseGroup = this.examChecklistService.checkEachGroupContainsExercise(this.exam);
        const maximumPointsEqual = this.examChecklistService.checkPointsExercisesEqual(this.exam);
        const examPointsReachable = this.examChecklistService.checkTotalPointsMandatory(maximumPointsEqual, this.exam);
```

Thanks for addressing my comment here.
If you would have objected to the suggested change I would have accepted that as well as the balance between enough documentation and too much doc is quite hard and depends a lot on individual preferences and experiences.

> In general I have the feeling that we are a bit economical when it comes to in-depth method documentation. This often leads to code that is hard to maintain and modify for any developer other than the original author.

That is my experience as well. There are definitely a lot of places where the documentation for non-trivial methods should be improved when touching the code.

willergeorg replied:
We do not come to a common denominator in terms of code structure, do we? 😄

b-fein reviewed (dismissed):
Thanks for the changes, I added some more comments inline. :+1:

b-fein reviewed (dismissed):
Thanks, code looks good now! :+1:

Malyuk-A reviewed (dismissed):
After discussing the changes with @willergeorg the code changes look good to me

filip-gregurevic reviewed (approved):
Changes look good to me, nice work 🚀

canberkanar commented:
![image](https://user-images.githubusercontent.com/67344158/152433774-f13c448f-3366-47d7-80c8-a123519921c1.png)

I have done manual testing and could not notice any unwanted behavior nor inconsistency between the developed feature and the Exam Checklist. 

One thing I saw is that even though it is consistent with what is done on the Exam Checklist before, I found Generate student exams: 0 of 0 to be confusing. 0 students have 0 exams and a check next to it. Maybe this part could be rephrased like exams to be generated?

willergeorg commented:
> ![image](https://user-images.githubusercontent.com/67344158/152433774-f13c448f-3366-47d7-80c8-a123519921c1.png)
> 
> I have done manual testing and could not notice any unwanted behavior nor inconsistency between the developed feature and the Exam Checklist.
> 
> One thing I saw is that even though it is consistent with what is done on the Exam Checklist before, I found Generate student exams: 0 of 0 to be confusing. 0 students have 0 exams and a check next to it. Maybe this part could be rephrased like exams to be generated?

Good catch! This is definitely confusing and should be handled differently. I changed the implementation of the roadmap that way, that for the case that no student is registered yet the exam generation step is formulated more conveniently.
![image](https://user-images.githubusercontent.com/80622272/152526877-8b99e0ac-4dfd-480d-b279-b8b0c45de028.png)
Let me know what you thing about it!

canberkanar commented:
@willergeorg looking good, nice job!

dveljacic reviewed (dismissed):
Tested on TS1. Didn't find any inconsistencies. Everything looks good. ✅

ole-ve reviewed (dismissed):
Tested on TS1, the described checks work as described. However I have some notes, that may be out of scope.

ole-ve reviewed (commented):
Tested on TS1, I have some notes:
1. Following configuration does not fulfill the check for the reachability of the maximum points:
<img width="711" alt="Bildschirmfoto 2022-02-05 um 18 00 00" src="https://user-images.githubusercontent.com/73833780/152652171-f4f5ded3-34be-4e4b-a9ee-51bd1d49c050.png">
Maximum number of points: 6, Number of Exercises: 3
I am not sure if this is an error or expected behavior (since it would be possible to reach this number of points with bonus points).

2. Most likely not related to this PR, but found while testing:
<img width="1403" alt="Bildschirmfoto 2022-02-05 um 18 19 28" src="https://user-images.githubusercontent.com/73833780/152652258-e1ca207b-e355-4c2f-b5d9-4341ae3b1610.png">
The student participated in the exercise and this was the student's review view. The console shows, that `course` is undefined when calling `course!.complaintsEnabled`.


3. After the review deadline has ended and refreshing the page, the console gives out the following error:
exam-participation-summary.component.ts (line 67): `this.studentExams.exercises` is not iterable


4. When clicking 'View result' for an exercise, the console gives out the following error:
header-participation-page.component.ts (line 56): `exercise.categories` with 'categories' being undefined.


While 3. and 4. seem to have no direct influence on the usability, 2. leads to students not being able to hand in complaints and only seeing their detailed result after the review deadline.

ole-ve reviewed (commented):
I couldn't reproduce the error when creating a new exam. I will have to take a deeper look into this.

willergeorg commented:
> Tested on TS1, I have some notes:
> 
> 1. Following configuration does not fulfill the check for the reachability of the maximum points:
> 
> <img alt="Bildschirmfoto 2022-02-05 um 18 00 00" width="711" src="https://user-images.githubusercontent.com/73833780/152652171-f4f5ded3-34be-4e4b-a9ee-51bd1d49c050.png">
> 
> Maximum number of points: 6, Number of Exercises: 3
> I am not sure if this is an error or expected behavior (since it would be possible to reach this number of points with bonus points).
> 
> 2. Most likely not related to this PR, but found while testing:
> 
> <img alt="Bildschirmfoto 2022-02-05 um 18 19 28" width="1403" src="https://user-images.githubusercontent.com/73833780/152652258-e1ca207b-e355-4c2f-b5d9-4341ae3b1610.png">
> 
> The student participated in the exercise and this was the student's review view. The console shows, that `course` is undefined when calling `course!.complaintsEnabled`.
> 3. After the review deadline has ended and refreshing the page, the console gives out the following error:
>    exam-participation-summary.component.ts (line 67): `this.studentExams.exercises` is not iterable
> 4. When clicking 'View result' for an exercise, the console gives out the following error:
>    header-participation-page.component.ts (line 56): `exercise.categories` with 'categories' being undefined.
> 
> While 3. and 4. seem to have no direct influence on the usability, 2. leads to students not being able to hand in complaints and only seeing their detailed result after the review deadline.

I discussed your configuration with @krusche as the underlying logic is legacy code and I do not have the knowledge to judge whether this works as intended or not. It seems to be the case that the legacy code works as expected and the bonus points for an exercise should not be taken into account for the mandatory points check.
Your remaining points, as you already stated, do not seem to be related to this PR. Although they might need further inspection as this situation must not occur on production. Please create an issue for that if you manage to reproduce the problem.

b-fein commented on src/main/webapp/app/exam/manage/exam-status.component.ts:
```diff
@@ -0,0 +1,172 @@
+    examChecklist: ExamChecklist;
+    numberOfGeneratedStudentExams: number;
+
+    configureExercises: boolean;
```
Nit: I think `configuredExercises` would be a better name as the variable shows that the instructor has already successfully finished the associated task. Applies to the next three variables as well.

willergeorg replied:
I chose the variable names so that they refer to the steps displayed to the user. These steps are formulated as action items on purpose. But I agree that it might be confusing in the code. Changed in 7f41df2

b-fein replied:
Thanks!

b-fein reviewed (dismissed):
Code changes since the last review look good. I only left one small inline comment about some variable names. :+1:

b-fein reviewed (approved):
Code: :+1:

Malyuk-A reviewed (approved):
I looked at the changes since my last approval. Looks good to me. Re-approve

filip-gregurevic reviewed (approved):
Code changes since my last approval look good to me 👍🏼
